typing/ikinds/ikinds.ml

```diff
@@ -0,0 +1,1009 @@
+  then
+    Jkind.sub_jkind_l ?allow_any_crossing ~type_equal ~context ~level sub super
+  else
+    let allow_any =
+      match allow_any_crossing with Some true -> true | _ -> false
+    in
+    if allow_any
+    then (
+      (if !Types.ikind_debug
+      then
+        let origin_suffix =
+          match origin with None -> "" | Some o -> " origin=" ^ o
+        in
+        Format.eprintf
+          "[ikind-subjkind] call%s allow_any=true@;sub=%a@;super=%a@."
+          origin_suffix Jkind.format sub Jkind.format super);
+      Ok ())
+    else
+      let ctx = make_ctx ~context in
+      let super_poly = ckind_of_jkind_l ctx super in
+      let super_is_constant =
+        Ldd.solve_pending ();
+        Ldd.is_const super_poly
+      in
+      let sub_ctx =
+        if super_is_constant
+        then JK.reset_for_mode ctx ~mode:JK.Round_up
+        else ctx
+      in
+      let sub_poly = ckind_of_jkind_l sub_ctx sub in
+      let violating_axes = Ldd.leq_with_reason sub_poly super_poly in
+      (if !Types.ikind_debug
+      then
+        let origin_suffix =
+          match origin with None -> "" | Some o -> " origin=" ^ o
+        in
+        Format.eprintf
+          "[ikind-subjkind] call%s allow_any=false@;\
+           sub=%a@;\
+           super=%a@;\
+           @;\
+           sub_poly=%s@;\
+           super_poly=%s@."
+          origin_suffix Jkind.format sub Jkind.format super (Ldd.pp sub_poly)
+          (Ldd.pp super_poly));
+      match violating_axes with
+      | [] -> Ok ()
+      | _ ->
+        let () =
+          if !Types.ikind_debug
+          then
+            let axes =
+              violating_axes
+              |> List.map Axis_lattice.axis_number_to_axis_packed
+              |> List.map (fun (Jkind_axis.Axis.Pack ax) ->
+                     Jkind_axis.Axis.name ax)
+              |> String.concat ", "
+            in
+            Format.eprintf
+              "[ikind-subjkind] failure on axes: %s@;sub=%a@;super=%a@." axes
+              Jkind.format sub Jkind.format super
+        in
+        (* Do not try to adjust allowances; Violation.Not_a_subjkind
+           accepts an r-jkind. *)
+        let axis_reasons =
+          List.map
+            (fun axis ->
+              let axis_name = Axis_lattice.axis_number_to_axis_packed axis in
+              Jkind.Sub_failure_reason.Axis_disagreement axis_name)
+            violating_axes
+        in
+        Error
+          (Jkind.Violation.of_ ~context
+             (Jkind.Violation.Not_a_subjkind (sub, super, axis_reasons)))
+
+let sub ?origin ~(type_equal : Types.type_expr -> Types.type_expr -> bool)
+    ~(context : Jkind.jkind_context) ~level (sub : Types.jkind_l)
+    (super : Types.jkind_r) : bool =
+  ignore origin;
+  if not !Clflags.ikinds
+  then Jkind.sub ~type_equal ~context ~level sub super
+  else
+    let ctx = make_ctx ~context in
+    match
+      JK.leq_with_reason (ckind_of_jkind_l ctx sub) (ckind_of_jkind_r super)
```

Comment:
Should this just use `leq` rather than `leq_with_reason`? Or maybe should we go further and just get rid of `leq` in favor of only having `leq_with_reason` (I'm guessing there's a perf penalty to doing this)?